### ISSUE TITLE
Failure running miner for GeForce 940M

### ISSUE THREAD

**Issue by Antokolos** [enerc/VulkanXMRMiner#3]
Hello!
My OS in Windows 10
My card is GeForce 940M (running on Acer Aspire E 15 laptop)
The card is correctly determined by the miner:

Checking your cards
Index:0 MemorySize:4 Gb - GeForce 940M

Finding best configuration....
Card:0  4 Compute Units/Stream Multiprocessors, using factor -144 (-576 threads), local size 16

(Please note, factor is negative, is this correct???)

When the miner starts mining, it is terminated with the following message:
Failure in vkBindBufferMemory at 236 ../src/mvulkan.cpp  ErrCode=-1000011001

**Comment by enerc** [comment]
Hello,

(You must be the first to test on Nvidia hardware and looks Nvidia drivers reports memory differently than Amd - They include CPU shared memory and this explains the negative numbers - I will fix it). 

Anyway, you can edit config.json file and start with:
 "cards" : [
  {
    "index"    : "0",
    "cu"       : "12",
    "factor"   : "32",
    "worksize" : "16"
  } ,
I think your card has 384 cores, so 12*32 should be a good start.

**Comment by Antokolos** [comment]
Tried 12-32-16 config, but unfortunately the result is the same (
Maybe I should install some additional drivers? Currently I can use this card to mine Monero using the xmrig-nvidia miner, but because of the Monero hardfork I decided to try something new :)
Here is the parameters I use for this card on xmrig-nvidia, maybe it can help:
"threads": [
        {
            "index": 0,
            "threads": 48,
            "blocks": 4,
            "bfactor": 6,
            "bsleep": 80,
            "affine_to_cpu": false
        }
    ],
![image](https://user-images.githubusercontent.com/7196697/53981292-6cc1e380-413c-11e9-8459-c9c853e458c5.png)

**Comment by enerc** [comment]
ErrCode=-1000011001 is VK_ERROR_VALIDATION_FAILED_EXT. Very helpful Vulkan error code...
Could you try "cu" : "1", "factor" : "1" to see if also fails. It can allocate memory but fails to map it.
What's your driver version?
Thanks a lot

**Comment by Antokolos** [comment]
Yes, it also fails with cu 1, factor 1
Driver is GeForce game ready driver 398.82 08/01/2018
I'll try to install the latest driver

**Comment by Antokolos** [comment]
Tried to run the miner with "cu" : "1", "factor" : "1" and driver version 419.35, but still with no luck.
Don't worry, though, I can wait for the updated version without any problems, hope that you can fix this problem in the future!

**Comment by enerc** [comment]
Could you please post Vulkaninfo.exe program output, especially the section like this one:

VkQueueFamilyProperties[0]:
        queueFlags         = GRAPHICS | COMPUTE | TRANSFER | SPARSE
        queueCount         = 1
        timestampValidBits = 64
        minImageTransferGranularity = (1, 1, 1)

VkQueueFamilyProperties[1]:
        queueFlags         = COMPUTE | TRANSFER | SPARSE
        queueCount         = 4
        timestampValidBits = 64
        minImageTransferGranularity = (1, 1, 1)

VkQueueFamilyProperties[2]:
        queueFlags         = TRANSFER | SPARSE
        queueCount         = 2
        timestampValidBits = 64
        minImageTransferGranularity = (16, 16, 8)


VkPhysicalDeviceMemoryProperties:
        memoryHeapCount       = 3
        memoryHeaps[0] :
                size          = 8304721920 (0x1ef000000) (7.73 GiB)
                flags:
                        VK_MEMORY_HEAP_DEVICE_LOCAL_BIT
        memoryHeaps[1] :
                size          = 8372879360 (0x1f3100000) (7.80 GiB)
                flags:
                        None
        memoryHeaps[2] :
                size          = 268435456 (0x10000000) (256.00 MiB)
                flags:
                        VK_MEMORY_HEAP_DEVICE_LOCAL_BIT
        memoryTypeCount       = 4
        memoryTypes[0] :
                heapIndex     = 0
                propertyFlags = 0x1:
                        VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT
 
Thanks for your help!

**Comment by Antokolos** [comment]
Here it is!
[vulkaninfo.zip](https://github.com/enerc/VulkanXMRMiner/files/2946160/vulkaninfo.zip)

**Comment by enerc** [comment]
I think I have found the issue: I missed this sentence in Vulkan specs "memoryOffset must be an integer multiple of the alignment member". Alignment is 4 on AMD and 0x10000 on the file you provided.
I have updated the repository and uploaded a build called vulkanxmrminer-latest.zip
Hope this will fix it!

**Comment by Antokolos** [comment]
It works, thank you very much!
![image](https://user-images.githubusercontent.com/7196697/54041364-bb7d8500-41f1-11e9-9fb9-4fb7df624469.png)

**Comment by BigslimVdub** [comment]
Seems this is not working in the latest .4 build for k12. Crashes on 399.** drivers when trying to run any settings on windows 10 when miner starts.

**Comment by enerc** [comment]
Hi BigSlim,
Could you please post Vulkaninfo.exe program output? What's the error reported by the miner when you run it on a cmd?

**Comment by BigslimVdub** [comment]
Error code -4

Failure in vkQueueSubmit at 366 ../src/miner.cpp

EDIT: added output for you. Updated to 430 game ready drivers and it seems to attempt to start but still fails. 
[VulkanInfoGT710.txt](https://github.com/enerc/VulkanXMRMiner/files/3140385/VulkanInfoGT710.txt)

**Comment by enerc** [comment]
It's the usual Windows TDR (Timeout Detection and Recovery) when a kernel last for more than 2 seconds (https://docs.microsoft.com/en-us/windows-hardware/drivers/display/timeout-detection-and-recovery).
To solve it:
- either you reduce Cu and Factor in the configuration file (config.json) for the CN kernel to take less than 2 seconds,
- or you apply the registry change provided in the dist zip file. This registry change removes TDR

**Comment by enerc** [comment]
Could be that it ran out of registers. Could you try to replace the kernel with the attached one? (unzip the file)
[k12_spv.zip](https://github.com/enerc/VulkanXMRMiner/files/3146834/k12_spv.zip)

**Comment by LonelyCodeCat** [comment]
Vulkan SPIR-V XMR Miner 0.4


If you experience very low hashrate, close apps like firefox, chrome,..., start the miner, and then reopen them.
Connecting to 162.210.173.150:7777 ... done!
Driver API version 1.1.99
Using XMR Variant: KangarooTwelve
Mining started.... (Press q to stop)
[Time] 'Total H/s' 'Good shares'/'Invalid shares'/'Expired shares' [GPU] H/s 'good hashes'/'bad hashes'
[00:00:00] 958.6MH/s 0/0/0 [0]:958.6Mh/s 0/0
[00:00:30] 963.6MH/s 0/0/0 [0]:963.6Mh/s 0/0
[00:01:00] 957.7MH/s 0/0/0 [0]:957.7Mh/s 0/0

**Comment by LonelyCodeCat** [comment]
[VULKANINFO.txt](https://github.com/enerc/VulkanXMRMiner/files/3148842/VULKANINFO.txt)

**Comment by enerc** [comment]
@857487  Thanks a lot for reporting! The results are from the 0.4 release, not with the file above, right? 
(the above patch file is for debugging and will report 4x more hashrate than actually processed).

The 0.4.1 will come with an improved kernel (at least on AMD). If you want to test it you can grab k12.spv file from github (master branch)  and replace it in your installation. I'am currious how it does on Nvidia HW - Was 40% faster on AMD.

**Comment by LonelyCodeCat** [comment]
@enerc
https://github.com/enerc/VulkanXMRMiner/files/3146834/k12_spv.zip
I can use the above file to replace it before I can run it, or I will get Error code -4
Failure in vkQueueSubmit at 366 ../src/miner.cpp, but even if k12_spv is replaced, it still doesn't work. Good shares is 0.

**Comment by enerc** [comment]
162.210.173.150 has an issue on port 7777 since it returns "target":"00000000" in the connecting string, meaning the difficulty is infinite.
I don't get any result on 7777 on this pool either.

Could you try on port 5555 ? Works for me (100k good results in 30 seconds)

**Comment by stoffu** [comment]
My card is GeForce GTX 980 and the above linked k12.spv was indeed necessary to run the miner (along with some additional fix #16).
Why not put it into the repository?

The pool 162.210.173.150:7777 returning target of 0 is now fixed.

```
Vulkan SPIR-V XMR Miner 0.4.1


If you experience very low hashrate, close apps like firefox, chrome,..., start the miner, and then reopen them.
Connecting to 162.210.173.150:7777 ... done!
SEND {"method":"login","params":{"login":"UmiFiLbUNYxd3AfnD4a8b5G7jqC4pjzLxWt6t958twa4bXw6KyairqHDdqn1d5Udu9CU59WBwZq8Z7yCVd8uBrZv2KbZgxLKH","pass":"x","rigid":"","agent":"vulkan XMR miner"},"id":1}
RECV {"id":1,"jsonrpc":"2.0","error":null,"result":{"id":"854044906282797","job":{"blob":"08089e9eb9e705dd316be6cbe8ae85667b54a92c21e447197005171fc7e753a8020a85f1bf1fdc0000000000000000dcb98ef606645976dbb3ae7ff458cb126850ec97f3163b028d9fc45744df5d1b01","job_id":"959785932977683","target":"f0bffe0a00000000"},"status":"OK"}}
Driver API version 1.1.99
Using XMR Variant: KangarooTwelve
Mining started.... (Press q to stop)
[Time] 'Total H/s' 'Good shares'/'Invalid shares'/'Expired shares' [GPU] H/s 'good hashes'/'bad hashes'
[13:42:52] 303.5MH/s 0/0/0 [0]:275.8Mh/s 0/0  [1]:1.7Mh/s 0/0  [2]:1.7Mh/s 0/0  [3]:1.7Mh/s 0/0  [4]:1.7Mh/s 0/0  [5]:1.7Mh/s 0/0  [6]:1.7Mh/s 0/0  [7]:1.7Mh/s 0/0  [8]:1.7Mh/s 0/0  [9]:1.7Mh/s 0/0  [10]:1.7Mh/s 0/0  [11]:1.7Mh/s 0/0  [12]:1.7Mh/s 0/0  [13]:1.7Mh/s 0/0  [14]:1.7Mh/s 0/0  [15]:1.7Mh/s 0/0  [16]:1.7Mh/s 0/0
[13:43:23] 253.8MH/s 0/0/0 [0]:226.0Mh/s 0/0  [1]:1.7Mh/s 0/0  [2]:1.7Mh/s 0/0  [3]:1.7Mh/s 0/0  [4]:1.7Mh/s 0/0  [5]:1.7Mh/s 0/0  [6]:1.7Mh/s 0/0  [7]:1.7Mh/s 0/0  [8]:1.7Mh/s 0/0  [9]:1.7Mh/s 0/0  [10]:1.7Mh/s 0/0  [11]:1.7Mh/s 0/0  [12]:1.7Mh/s 0/0  [13]:1.7Mh/s 0/0  [14]:1.7Mh/s 0/0  [15]:1.7Mh/s 0/0  [16]:1.7Mh/s 0/0
[13:43:54] 220.8MH/s 0/0/0 [0]:193.1Mh/s 0/0  [1]:1.7Mh/s 0/0  [2]:1.7Mh/s 0/0  [3]:1.7Mh/s 0/0  [4]:1.7Mh/s 0/0  [5]:1.8Mh/s 0/0  [6]:1.7Mh/s 0/0  [7]:1.7Mh/s 0/0  [8]:1.7Mh/s 0/0  [9]:1.8Mh/s 0/0  [10]:1.7Mh/s 0/0  [11]:1.7Mh/s 0/0  [12]:1.7Mh/s 0/0  [13]:1.7Mh/s 0/0  [14]:1.7Mh/s 0/0  [15]:1.7Mh/s 0/0  [16]:1.7Mh/s 0/0
[13:44:25] 299.3MH/s 0/0/0 [0]:271.6Mh/s 0/0  [1]:1.7Mh/s 0/0  [2]:1.7Mh/s 0/0  [3]:1.7Mh/s 0/0  [4]:1.7Mh/s 0/0  [5]:1.7Mh/s 0/0  [6]:1.7Mh/s 0/0  [7]:1.7Mh/s 0/0  [8]:1.7Mh/s 0/0  [9]:1.7Mh/s 0/0  [10]:1.8Mh/s 0/0  [11]:1.7Mh/s 0/0  [12]:1.7Mh/s 0/0  [13]:1.7Mh/s 0/0  [14]:1.7Mh/s 0/0  [15]:1.7Mh/s 0/0  [16]:1.7Mh/s 0/0
[13:44:56] 234.3MH/s 0/0/0 [0]:206.6Mh/s 0/0  [1]:1.7Mh/s 0/0  [2]:1.7Mh/s 0/0  [3]:1.7Mh/s 0/0  [4]:1.7Mh/s 0/0  [5]:1.7Mh/s 0/0  [6]:1.7Mh/s 0/0  [7]:1.7Mh/s 0/0  [8]:1.7Mh/s 0/0  [9]:1.7Mh/s 0/0  [10]:1.7Mh/s 0/0  [11]:1.7Mh/s 0/0  [12]:1.7Mh/s 0/0  [13]:1.7Mh/s 0/0  [14]:1.7Mh/s 0/0  [15]:1.7Mh/s 0/0  [16]:1.7Mh/s 0/0
[13:45:26] 215.8MH/s 0/0/0 [0]:188.0Mh/s 0/0  [1]:1.7Mh/s 0/0  [2]:1.7Mh/s 0/0  [3]:1.7Mh/s 0/0  [4]:1.7Mh/s 0/0  [5]:1.7Mh/s 0/0  [6]:1.7Mh/s 0/0  [7]:1.7Mh/s 0/0  [8]:1.7Mh/s 0/0  [9]:1.7Mh/s 0/0  [10]:1.8Mh/s 0/0  [11]:1.7Mh/s 0/0  [12]:1.7Mh/s 0/0  [13]:1.7Mh/s 0/0  [14]:1.7Mh/s 0/0  [15]:1.7Mh/s 0/0  [16]:1.7Mh/s 0/0
```

**Comment by enerc** [comment]
I have created one kernel (spirv file) for AMD and another with less registers for Nvidia on master.
I hope HW detection will work properly to choose the right kernel.
Hashrate on Nvidia HW should now be correct and no patches should be needed (provided my asumption of 1/2 less registers works - used to be 1/4 in the above patch)

**Comment by enerc** [comment]
Nvidia fixes added in 0.4.1 release
